### PR TITLE
Feat/cs/app messages

### PR DESCRIPTION
**Pull request by ReagentX**
Implement App Messages!

- Parse third party `MSMessageTemplateLayout` message payloads
- Parse first party custom message payloads

### CODE REVIEW

File: imessage-database/src/util/plist.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+///
+/// For example, a document with a root pointing to some XML like
+///
+/// ```xml
+/// <array>
+///     <dict>
+///         <key>link</key>
+///         <dict>
+///              <key>CF$UID</key>
+///              <integer>47</integer>
```
Comment (ReagentX) [inline imessage-database/src/util/plist.rs]: `47` should be `2`

File: imessage-database/src/message_types/url.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
+/*!
+These are the link previews that iMessage generates when sending links and can
```
Comment (ReagentX) [inline imessage-database/src/message_types/url.rs]: This wording is awkward

File: imessage-exporter/src/exporters/html.rs [hunk 3]
```diff
@@ -384,17 +407,65 @@ impl<'a> Writer<'a> for HTML<'a> {
...
+                            Err(why) => return Err(PlistParseError::ParseError(format!("{why}"))),
+                        }
+                    };
+                    app_bubble.push_str(&res);
+                }
+                None => {
+                    // Sometimes, URL messages are missing their payloads
+                    if message.is_url() {
+                        if let Some(text) = &message.text {
+                            return Ok(text.to_string());
```
Comment (ReagentX) [inline imessage-exporter/src/exporters/html.rs]: This should be a hyperlink, not just the URL text

File: imessage-exporter/src/exporters/html.rs [hunk 4]
```diff
@@ -215,12 +230,20 @@ impl<'a> Writer<'a> for HTML<'a> {
                     }
                 }
                 // TODO: Support app messages
```
Comment (ReagentX) [inline imessage-exporter/src/exporters/html.rs]: Remove todo!

File: imessage-exporter/src/exporters/txt.rs [hunk 5]
```diff
@@ -190,7 +209,7 @@ impl<'a> Writer<'a> for TXT<'a> {
             );
         }
 
-        // TODO: We add 2 newlines for messages that have replies
+        // TODO: We add 2 newlines for messages that have replies with `has_replies`
```
Comment (ReagentX) [inline imessage-exporter/src/exporters/txt.rs]: Address this

File: docs/binary/features.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+This tool targets the current latest public release for MacOS and iMessage.
+
+## Supported Message Features
+
+- Plain Text
+  - Correctly extracts time-zone corrected timestamps
+  - Detects when a message was read and calculates the time until read for both parties
+  - Detects the service a message was sent from
+    - In HTML exports, balloons are colored correctly for the service they were sent with
+- Mutli-part messages
```
Comment (ebramanti) [inline docs/binary/features.md]:
```suggestion
- Multi-part messages
```

File: imessage-exporter/src/exporters/txt.rs (hunk 5 above)
Comment (ReagentX) [reply imessage-exporter/src/exporters/txt.rs]: Honestly, the more I read the exports the more I like the extra space

File: imessage-database/src/message_types/app.rs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+            title: Some("Rusty Grill - Boise"),
+            subtitle: Some("Reservation Confirmed"),
+            caption: Some("Table for 4 people\nSunday, October 17 at 7:45 PM"),
+            subcaption: Some("You're invited! Tap to accept."),
+            trailing_caption: None,
+            trailing_subcaption: None,
+            app_name: Some("OpenTable"),
+            ldtext: None,
+        };
+        println!("{:?}", balloon);
```
Comment (ebramanti) [inline imessage-database/src/message_types/app.rs]: Do you still need this? Same with below functions

File: imessage-database/src/message_types/music.rs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    #[test]
+    fn test_parse_apple_music() {
+        let plist_path = current_dir()
+            .unwrap()
+            .as_path()
+            .join("test_data/AppleMusic.plist");
+        let plist_data = File::open(plist_path).unwrap();
+        let plist = Value::from_reader(plist_data).unwrap();
+        let parsed = parse_plist(&plist).unwrap();
+        println!("{parsed:?}");
```
Comment (ebramanti) [inline imessage-database/src/message_types/music.rs]: Same question here

File: imessage-exporter/src/app/converter.rs [hunk 9]
```diff
@@ -6,21 +6,13 @@ use std::{
...
-/// Docs: https://www.unix.com/man-page/osx/1/sips/ (or `man sips`)
+/// Docs: <https://www.unix.com/man-page/osx/1/sips/> (or `man sips`)
 pub fn heic_to_jpeg(from: &Path, to: &Path) -> Option<()> {
     // Get the path we want to copy from
-    let from_path = if let Some(from_path) = from.to_str() {
-        from_path
-    } else {
-        return None;
-    };
+    let from_path = from.to_str()?;
```
Comment (ebramanti) [inline imessage-exporter/src/app/converter.rs:12]: Nice one

File: imessage-database/src/util/plist.rs [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                    }
+                }
+            }
+            Ok(plist::Value::Dictionary(dictionary))
+        }
+        Value::Uid(uid) => {
+            follow_uid(objects, uid.get() as usize, None, None)
+        }
+        Value::Boolean(_) | Value::Integer(_) | Value::String(_) => Ok(item.to_owned()),
+        _ => Ok(item.to_owned()),
```
Comment (ebramanti) [inline imessage-database/src/util/plist.rs]: Can this just be a catch-all? Or do you prefer to enumerate these base types for clarity?

File: imessage-database/src/util/plist.rs [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    }
+}
+
+pub fn extract_parsed_str<'a>(
+    payload: &'a HashMap<&'a str, &'a Value>,
+    key: &str,
+) -> Option<&'a str> {
+    payload
+        .get(key)
+        .and_then(|item| item.as_string())
```
Comment (ebramanti) [inline imessage-database/src/util/plist.rs:197]: why can't `as_string` be called directly here?

File: imessage-exporter/src/exporters/html.rs [hunk 12]
```diff
@@ -181,14 +190,20 @@ impl<'a> Writer<'a> for HTML<'a> {
             );
 
             match message_part {
-                BubbleType::Text(text) => {
-                    self.add_line(
+                BubbleType::Text(text) => match text.starts_with(FITNESS_RECEIVER) {
```
Comment (ebramanti) [inline imessage-exporter/src/exporters/html.rs]: Any reason to use `match` over `if/else` here?

File: imessage-exporter/src/exporters/resources/example.html [hunk 13]
```diff
@@ -19,6 +19,28 @@
     </div>
 </div>
 
+<div class="message">
+    <div class="sent iMessage">
+        <p><span class="timestamp">Jul 17, 2022 5:04:05 PM</span>
```
Comment (ebramanti) [inline imessage-exporter/src/exporters/resources/example.html:24]: Based 🎂

**Review by ebramanti — DISMISSED** [review]

File: imessage-database/src/util/plist.rs (hunk 11 above)
Comment (ReagentX) [reply imessage-database/src/util/plist.rs:197]: `.get(key)` returns `Option<Value>`, so `and()` handles the `None`.

File: imessage-exporter/src/exporters/html.rs (hunk 12 above)
Comment (ReagentX) [reply imessage-exporter/src/exporters/html.rs]: I thought match was cleaner than having the lines with braces, but it does seem odd to `match` on `bool`. I actually do this in a few places; I need to think about the style here.

Comment (ReagentX) [reply imessage-exporter/src/exporters/html.rs]: Which style do you think is cleaner? 

```rust
match text.starts_with("https://music.apple") {
    true => Variant::App(CustomBalloon::Music),
    false => Variant::App(CustomBalloon::URL),
}
```

or 

```rust
if text.starts_with("https://music.apple") {
    Variant::App(CustomBalloon::Music)
} else {
    Variant::App(CustomBalloon::URL)
}
```

I tend to lean towards pattern matching here. Maybe there is a way to do this with a method chain...

Comment (ReagentX) [reply imessage-exporter/src/exporters/html.rs]: This works:

```rust
text.starts_with("https://music.apple")
    .then_some(Variant::App(CustomBalloon::Music))
    .unwrap_or(Variant::App(CustomBalloon::URL))
```

But `clippy` doesn't like it:

```
warning: use of `.then_some(..).unwrap_or(..)` can be written more clearly with `if .. else ..`
   --> imessage-database/src/tables/messages.rs:570:29
    |
570 | / ...                   text.starts_with("https://music.apple")
571 | | ...                       .then_some(Variant::App(CustomBalloon::Music))
572 | | ...                       .unwrap_or(Variant::App(CustomBalloon::URL))
    | |______________________________________________________________________^ help: try: `if text.starts_with("https://music.apple") { Variant::App(CustomBalloon::Music) } else { Variant::App(CustomBalloon::URL) }`
    |
    = note: `#[warn(clippy::obfuscated_if_else)]` on by default
```

**Pull request merged**